[Merged by Bors] - feat(tactic/linarith): nlinarith tactic

digama0 opened the pull request:
Based on Coq's [nra](https://coq.inria.fr/refman/addendum/micromega.html#coq:tacn.nra) tactic, and requested on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/nonlinear.20linarith).

digama0 committed:
feat(tactic/linarith): nra tactic

bryangingechen commented on src/tactic/linarith.lean:
```diff
@@ -895,4 +900,66 @@ add_tactic_doc
+  the context.
+* For every pair of hypotheses `0 ≤ a`, `0 ≤ b` in the context, the assumption `0 ≤ a * b` is added
+  to the context (non-recursively).
+-/
```
```suggestion
```
If there's only one declaration named in `decl_names`, its docstring will automatically be used.

digama0 committed:
fix docs

PatrickMassot commented on src/tactic/linarith.lean:
```diff
@@ -895,4 +900,55 @@ add_tactic_doc
+  (s, ge0) ← (list.mfoldr (λ h ⟨s, l⟩, do
+      t ← infer_type h,
+      s ← find_squares s t,
+      return (s, match t with `(0 ≤ %%a) := h :: l | _ := l end))
```
Is there any reason to only handle `(0 ≤ %%a)` and not strict inequalities?

digama0 replied:
I don't know, I was just following the description. I think this is quite limited, for example what about general inequalities, don't we want to know `a - b` is nonnegative? If `linarith` is doing context normalization then this pass should probably come afterward, so that it isn't tripped up by a proof of `a >= 0` for example.

PatrickMassot replied:
@robertylewis any opinion?

robertylewis replied:
`linarith` is old code that's been extended a bunch of times by various feature requests. I think it would take a nontrivial refactor to adapt the normalization routines to be used here too. Probably the whole preprocessing could just be rewritten.

This is also irrelevant to handling strict inequalities. I've never used `nlinarith` in Coq, but the description really does sound like `linarith` + epsilon. @PatrickMassot you're asking for `linarith` + 2*epsilon. Shouldn't take much extra work to handle >=, <, and so on, so that's fair enough. But I think it's not worth any effort to add features beyond that.

IMO this should be considered a "cheer if it works, don't complain if it doesn't" tactic. There's an endless list of possible "but what if..." feature requests, each one allowing it to solve one more problem.

digama0 replied:
Indeed. I will remind you (@PatrickMassot) that when `nra` was proposed this was the *very first question* I asked about the tactic: what is the extent of what it should be able to do? If we grow it incrementally it will become a hodgepodge (as `linarith` is already, to some extent). I will add support for `0 < a` and `0 <= a` hypotheses, but that's all. Anything more than that should come with an analysis of why it's actually useful and how it contributes to the overall structure, because nonlinear arithmetic, being undecidable, is never going to have a satisfactory tactic covering everything that fits within its nominal scope.

robertylewis replied:
> because nonlinear arithmetic, being undecidable

Maybe we're thinking different things by "nonlinear arithmetic" here. The theory of real closed fields is decidable, which is an extension of what a "nonlinear equivalent of linarith" would do. But I don't expect to see a CAD implementation in Lean 3. I agree with you in spirit!

PatrickMassot replied:
I asked Heather to give some examples that she found useful in Coq.

digama0 committed:
add products of positive and nonnegative values

robertylewis commented on src/tactic/linarith.lean:
```diff
@@ -895,4 +900,66 @@ add_tactic_doc
+  the context.
+* For every pair of hypotheses `0 ≤ a`, `0 ≤ b` in the context, the assumption `0 ≤ a * b` is added
+  to the context (non-recursively).
+-/
```
```suggestion
* For every pair of hypotheses `0 < a`, `0 < b` in the context, the assumption `0 < a * b` is added
  to the context (non-recursively).
-/
```

robertylewis reviewed (commented):
@digama0 could you add some tests? Or maybe the people who requested this could suggest some tests? Otherwise it looks good to me!

robertylewis committed:
linarith improvement

robertylewis committed:
comments

robertylewis committed:
add tests

robertylewis committed:
improve factor cancellation inside multiplication

robertylewis committed:
remove exit

robertylewis committed:
cleanup

robertylewis committed:
store comps as sorted lists instead of rb_maps

robertylewis committed:
add test

robertylewis committed:
add docs

robertylewis committed:
add docs for real

robertylewis committed:
missed comments

robertylewis committed:
Merge branch 'master' into linarith-speedup

robertylewis committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

robertylewis committed:
fix doc string

robertylewis committed:
Apply suggestions from code review

Co-authored-by: Gabriel Ebner <gebner@gebner.org>

robertylewis committed:
remove unneeded code

robertylewis committed:
Update src/tactic/linarith.lean

Co-authored-by: Gabriel Ebner <gebner@gebner.org>

robertylewis committed:
remove use of cmp

robertylewis committed:
avoid repeating comparisons between lists

robertylewis committed:
move and doc of_list_core

robertylewis committed:
Merge branch 'linarith-speedup' into nra

robertylewis committed:
add test

robertylewis committed:
rearrange hypotheses before nlinarith preprocessing

This repeats work that linarith does for its own purposes.
Once the linarith frontend is refactored, this can be combined.
Right now things are tied together too intricately.

robertylewis committed:
handle some of Kevin's examples

robertylewis committed:
avoid symmetric comparisons

This avoids feeding some extra hypotheses to linarith,
by mapping over the upper diagonal of a product of lists
instead of the full product.
Ultimately it doesn't matter, as linarith will filter out the redundant comparisons.
But this is better for performance,
and if linarith changes to use lists instead of rb_sets,
these will no longer be filtered.

robertylewis committed:
merge master

bryangingechen commented on test/linarith.lean:
```diff
@@ -170,6 +170,7 @@ example (x y : ℚ) (h : x < y) : x ≠ y := by linarith
 
 example (x y : ℚ) (h : x < y) : ¬ x = y := by linarith
 
+<<<<<<< HEAD
```
whoops!

robertylewis committed:
bad merge

robertylewis committed:
update docs

robertylewis committed:
more docs

robertylewis committed:
linter

robertylewis committed:
move instance to bool file

hrmacbeth commented on src/tactic/linarith.lean:
```diff
@@ -1052,4 +1073,69 @@ add_tactic_doc
 
+/--
+An extension of `linarith` with some preprocessing to allow it to solve some nonlinear arithmetic
+problems. (Based on Coq's `nlinarith` tactic.) See `linarith` for the available syntax of options,
```
"Coq's `nra` tactic"

robertylewis replied:
```suggestion
problems. (Based on Coq's `nra` tactic.) See `linarith` for the available syntax of options,
```

robertylewis committed:
Update src/tactic/linarith.lean

digama0 commented on src/data/bool.lean:
```diff
@@ -145,3 +145,11 @@ lemma bxor_iff_ne : ∀ {x y : bool}, bxor x y = tt ↔ x ≠ y := dec_trivial
 
 end bool
+
+instance : decidable_linear_order bool :=
```
Can this not be a global instance? This seems a bit weird, I don't know many languages other than C where you can compare booleans (and in that case it is only because C is too weakly typed and thinks they are integers). Also a better proposition for this would be `a -> b`.

robertylewis replied:
Sure, I don't mind, although I don't think it does any harm. I was surprised that there was no lattice instance already that would give this comparison.

gebner replied:
I agree with Rob that `bool` should have an order, in particular `bool` should be a `complete_boolean_algebra`.  Even Haskell has an order on `Bool`: https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Ord.html#i:ic:Ord:Ord:1

robertylewis committed:
mention nlinarith in linarith docs
